### PR TITLE
Add an admin page for adding urls

### LINKED ISSUE #919: Add an admin page for adding sites to Checkmate's allow-list
## Context

Hypothesis runs an app called Via (https://via.hypothes.is/) that enables users to annotate arbitrary web pages: you visit https://via.hypothes.is/, paste in the URL of a page you'd like to annotate, and Via proxies the given URL and injects the Hypothesis annotation client into it.

Of course, Via could be abused to proxy malware, phishing, etc. Via therefore has an **allow-list** of safe sites that're allowed to be proxied. If a user tries to proxy a site that isn't on the allow-list they'll be presented with a **Content cannot be annotated yet** page that includes a link to a [form](https://web.hypothes.is/via-request/) where they can submit a request for the site to be added to the allow-list. [Example](https://via.hypothes.is/https://blahblahblah.com/):

Image

## Checkmate

The allow-list actually exists in a separate Hypothesis app called Checkmate: <https://github.com/hypothesis/checkmate>. Checkmate provides an API that Via calls to ask whether a given URL is allowed or not.

## Problem

When we receive requests from users to add sites to the allow-list, our internal process for adding the sites is time-consuming and cumbersome. This process wastes the time of Hypothesis support staff and developers, and means that our users have to wait a long time for their requests to annotate sites to be granted.

1. Support staff evaluate the requested site and, if it looks safe, they add it to the [Via allow requests spreadsheet](https://docs.google.com/spreadsheets/d/1g82noNwqN8Wzv3CplB_i4YP4iy9F_mhaSRz5xLUgfMY/)
2. Once they've collected a number of sites in the spreadsheet support post a request into Slack, asking developers to add the new sites to Checkmate's allow-list. [Example Slack post](https://hypothes-is.slack.com/archives/C1MA4E9B9/p1732650736108259).
3. A developer then:

   1. Logs in to Checkmate's admin pages at https://checkmate.hypothes.is/admin
   2. The admin pages print out a long command for the developer to run locally. The printed command runs an [add_to_allow_list.py](https://github.com/hypothesis/checkmate/blob/ca0a9201bded2e092b55f4a0f0918f5dc10dff29/bin/add_to_allow_list.py) script from the Checkmate repo, and includes a session token as a command line argument. The script uses the session token to make calls to a [Checkmate API](https://github.com/hypothesis/checkmate/blob/a61e85e214ae8762884af76767dc64fb540ef718/checkmate/views/ui/api/add_to_allow_list.py) to add the sites to the allow-list.
   3. The developer checks out Checkmate locally, downloads the spreadsheet locally as CSV, copy-pastes the long command from Checkmate's admin pages and runs the command locally to import the URLs from the CSV into Checkmate's allow-list

This process is documented in two Stack Overflow For Teams posts:

1. [What is the process for responding to Via allow-list requests?](https://stackoverflowteams.com/c/hypothesis/questions/444)
2. [How can I add a URL to the allow list?](https://stackoverflowteams.com/c/hypothesis/questions/448)

## Solution

Support staff can already log in to Checkmate's admin pages at https://checkmate.hypothes.is/admin. There should be an admin page here that allows support to paste a URL into a text field and click a <kbd>Add URL to allow-list</kbd> button to add the URL to Checkmate's allow-list.

Support staff should be able to do this entirely on their own using the admin pages, no developers should be involved.

## Done when

- [x] Hypothesis support staff can log in to Checkmate's admin pages at https://checkmate.hypothes.is/admin. (I think this is already done -- @seanh)
- [ ] The admin pages have a simple form where the user can paste in a URL, click a button to submit the form, and the URL will be added to the allow-list. The view that receives the HTTP form submission should do something equivalent to what the [add_to_allow_list API view](https://github.com/hypothesis/checkmate/blob/a61e85e214ae8762884af76767dc64fb540ef718/checkmate/views/ui/api/add_to_allow_list.py) currently does.
- [ ] The admin pages no longer print out a long command that you're supposed to run locally, since this will no longer be needed. 
- [ ] The `bin/add_to_allow_list.py` script has been removed.
- [ ] Checkmate's `add_to_allow_list` API (`POST https://checkmate.hypothes.is/api/rule`) has been removed: this will no longer be needed since the admin pages will instead be using an HTTP form submission to add rules
- [ ] The documentation on Stack Overflow For Teams has been updated. [How can I add a URL to the allow list?](https://stackoverflowteams.com/c/hypothesis/questions/448) can simply be deleted. [What is the process for responding to Via allow-list requests?](https://stackoverflowteams.com/c/hypothesis/questions/444) should be updated to reflect the new process where support staff can use the admin pages to add sites themselves.

### PR DESCRIPTION
**Pull request by mtomilov**
Fixes #919 

This PR only adds a form to add new urls without removing the old api and script.

Testing
-------

1. Log in as devdata_admin (password: pass).

2. Go to `http://localhost:9099/admin/` which will redirect to `/admin/allow_rule/`.

3. Type in a URL (e.g. `http://example.com`) and click the "Submit" button to add it to the list of allowed URLs.

4. Type in the same URL again and click the "Submit" button to see an error message.

### CODE REVIEW

File: checkmate/views/ui/admin.py [hunk 1]
```diff
@@ -24,3 +41,26 @@ def get(self):
...
+    @view_config(request_method="GET")
+    def get(self):
+        return {}
+
+    @view_config(request_method="POST")
+    def post(self):
+        url = self.request.params["url"]
+        try:
+            allow_rule = self.request.find_service(RuleService).add_to_allow_list(url)
+        except ResourceConflict as e:
```
Comment (mtomilov) [inline checkmate/views/ui/admin.py:64]: Reusing the same `ResourceConflict` here from the api version for now even considering its parent being `JSONAPIException`.

File: tests/functional/checkmate/views/ui/api/add_to_allow_list_test.py [hunk 2]
```diff
@@ -22,7 +22,7 @@ def test_a_request_with_a_session_cookie_succeeds(self, app, json_body):
                     "rule": "example.com/",
                     "tags": ["manual"],
                 },
-                "id": 1,
+                "id": response.json["data"]["id"],
```
Comment (mtomilov) [inline tests/functional/checkmate/views/ui/api/add_to_allow_list_test.py:25]: Now that we add multiple rules in functional tests we can't rely on `id` here being constant.

File: checkmate/views/ui/admin.py [hunk 3]
```diff
@@ -1,12 +1,29 @@
...
+from checkmate.services import RuleService
 
 
-@view_defaults(route_name="admin_pages", request_method="GET")
+@view_config(route_name="admin.index")
+def index(request):
+    return HTTPFound(location=request.route_url("admin.allow_rule"))
+
+
+@notfound_view_config(path_info="/admin/*", append_slash=True)
```
Comment (mtomilov) [inline checkmate/views/ui/admin.py:21]: Wanted to make admin work with or without a slash for convenience.

File: checkmate/routes.py [hunk 4]
```diff
@@ -13,7 +13,10 @@ def add_routes(config):
     config.add_static_view("static", "static/static", cache_max_age=3600)
 
     config.add_route("present_block", "/block")
-    config.add_route("admin_pages", "/admin")
+
+    config.add_route("admin.index", "/admin/")
+    config.add_route("admin.pages", "/admin/pages/")
```
Comment (mtomilov) [inline checkmate/routes.py:18]: Leaving it as is for now and will remove in subsequent PR.

File: checkmate/templates/admin/allow_rule.html.jinja2 [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+{% endif %}
+
+{% if allow_rule %}
+  <p>Allowed as {{ allow_rule.rule }} with hash {{ allow_rule.hash }}</p>
+{% endif %}
+
+<form action="{{ request.route_url('admin.allow_rule') }}" method="post">
+  <input type="hidden" name="csrf_token" value="{{ get_csrf_token() }}">
+  <label>
+    <input type="url" name="url" placeholder="URL" size="100">
```
Comment (mtomilov) [inline checkmate/templates/admin/allow_rule.html.jinja2:21]: Made it a bit longer for visibility.

File: checkmate/templates/admin/allow_rule.html.jinja2 [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  {% if messages %}
+    <div>
+      {% for message in messages %}
+        <p>{{ message.detail }}</p>
+      {% endfor %}
+    </div>
+  {% endif %}
+
+  {% if allow_rule %}
+    <p>Allowed as {{ allow_rule.rule }} with hash {{ allow_rule.hash }}</p>
```
Comment (mtomilov) [inline checkmate/templates/admin/allow_rule.html.jinja2:15]: Sticking to the same format we had in [Checkmate.allow_url](https://github.com/hypothesis/checkmate/blob/ca0a9201bded2e092b55f4a0f0918f5dc10dff29/bin/add_to_allow_list.py#L111).

File: checkmate/templates/admin/base.html.jinja2 [hunk 7]
```diff
@@ -0,0 +1,36 @@
+<!doctype html>
+<html lang="en">
+<head>
+  <title>Checkmate Admin</title>
+  <style>
```
Comment (marcospri) [inline checkmate/templates/admin/base.html.jinja2:5]: I reckon we should use bulma here like the LMS admin pages:

https://bulma.io/

I don't think we need to go overboard with styling, just slap the CSS import and tweak the HTML a bit to render it a bit nicely


See the base file in LMS:  https://github.com/hypothesis/lms/blob/main/lms/templates/admin/base.html.jinja2#L34


I'd do in a follow up PR.

**Review by marcospri — APPROVED** [review]
This is great, we should have done it a long time ago, nice :+1:

File: checkmate/templates/admin/base.html.jinja2 (hunk 7 above)
Comment (mtomilov) [reply checkmate/templates/admin/base.html.jinja2:5]: Sounds good no problem!

**Pull request merged**